src/util/fifo.h

```diff
@@ -67,3 +82,162 @@ class FIFO {
+    // to a region of count values. If less than count values
+    // are available, limit to the available amount.
+    // If the region surpasses the end of the ring buffer,
+    // the first region will be until the end, the second
+    // region will be the remainder from the start, otherwise
+    // the second region will be nullptr and size 0.
+    int aquireReadRegions(size_type count,
+            DataType** dataPtr1,
+            ring_buffer_size_t* sizePtr1,
+            DataType** dataPtr2,
+            ring_buffer_size_t* sizePtr2) {
+        size_type readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const size_type writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        const size_type available = writeIndex - readIndex;
+        count = std::min(available, count);
+        readIndex = readIndex & m_mask;
+        *sizePtr1 = static_cast<int>(std::min(m_size - readIndex, count));
+        *sizePtr2 = static_cast<int>(count) - *sizePtr1;
+        *dataPtr1 = m_data.data() + readIndex;
+        *dataPtr2 = *sizePtr2 == 0 ? nullptr : m_data.data();
+
+        return static_cast<int>(count);
+    }
+    // Advance the read index after aquireReadRegions. Count should
+    // be the return value of aquireReadRegions, which is the total
+    // region size acquired.
+    int releaseReadRegions(size_type count) {
+        const size_type readIndex = (m_readIndex.load(std::memory_order_relaxed) + count);
+        m_readIndex.store(readIndex, std::memory_order_release);
+        return static_cast<int>(readIndex & m_mask);
+    }
+    // Same as aquireReadRegions, for write operations
+    int aquireWriteRegions(size_type count,
+            DataType** dataPtr1,
+            ring_buffer_size_t* sizePtr1,
+            DataType** dataPtr2,
+            ring_buffer_size_t* sizePtr2) {
+        const size_type readIndex = m_readIndex.load(std::memory_order_acquire);
+        size_type writeIndex = m_writeIndex.load(std::memory_order_relaxed);
+        const size_type available = m_size - (writeIndex - readIndex);
+        count = std::min(available, count);
+        writeIndex = writeIndex & m_mask;
+        *sizePtr1 = static_cast<int>(std::min(m_size - writeIndex, count));
+        *sizePtr2 = static_cast<int>(count) - *sizePtr1;
+        *dataPtr1 = m_data.data() + writeIndex;
+        *dataPtr2 = *sizePtr2 == 0 ? nullptr : m_data.data();
+
+        return static_cast<int>(count);
+    }
+    // Same as releaseReadRegions, for write operations
+    int releaseWriteRegions(size_type count) {
+        const size_type writeIndex = (m_writeIndex.load(std::memory_order_relaxed) + count);
+        m_writeIndex.store(writeIndex, std::memory_order_release);
+        return static_cast<int>(writeIndex & m_mask);
+    }
+    // Advance the read index with count values, or maximum until the write index.
+    // Returns the new read index (wrapped inside the buffer size)
+    int flushReadData(size_type count) {
```

Comment:
Is this function allowed from reader, writer or both? Do we need a second one for the other counterpart?

Reply:
Only allowed from reader, but apparently it is only used once (in ./engine/sidechain/shoutconnection.cpp). We do not need a second one for the counter part and in fact we might as well remove this one, and simply call releaseReadRegions there, which would amount to the same.

And releaseReadRegions and releaseWriteRegions would be more aptly named advanceReadIndex, advanceWriteIndex. But all of that is IMO beyond the PR.

Reply:
Yes, so documenting this as a source comment is sufficient.